tripal/tripal
# Entity Publishing

# Tripal 4 Core Dev Task

Issue #1515 


### Tripal Version:  4.x

## Description



This PR introduces basic publishing to Tripal 4. Currently, a publish job submitted in this PR will:

- Create new Tripal Entities of the requested content type that do not already exist.
- Integrate data added to Chado outside of content forms (i.e. via importers or pre-existing) for a SUBSET OF FIELDS (see below)
- It works both for (1) content page created through UI then new data added to chado later and (2) data added and no entity exists yet.
- It can and should be re-run when new data is added to existing content pages outside of the UI.

**NOTE: Publish in Tripal 4 is meant to create entities (same as Tripal 3) but ALSO to sync data in chado with fields (new in Tripal 4). Data added to Chado outside the content add/edit form will not show up in fields (i.e. on the content page) without it being published.**

## Caveats

It has currently only been tested for feature-based content types although it will be tested with other content types before it is elevated above DRAFT status.

It will only work for fields where all store properties are on the base table. It DOES NOT yet work for linker tables, property fields, etc. This will be added in a future PR.

It only works for the following content types: 
- Gene: 50
- mRNA: 75
- Organism: 6
- Contact: 7
- Analysis: 7 (two are of type Genome Assembly and 2 are of type genome annotation)
- Project: 2 (both of type Genome Project)
- Protocol: 3
**NOTE: it should also work for the germplasm types but I haven't tested them yet.**

## Testing?




### Automated Testing

Publish Form Test
- Complete. Tests the form using a mock storage backend, ensures that the backend can add form elements to the form, checks form validation and submit.

FindValues() Test
- Complete and working for the base table fields. Tests publishing of 50 genes.

### Manual Testing

1. Prepare a site with test data [according to this GIST](https://gist.github.com/laceysanderson/7c57bd173d354e432eb83f12360936b6).
3. Go to Admin > Tripal > Content and click on "Publish Tripal Content"
4. Choose "Chado Storage" as the data backend and then select a content type. Click publish to submit the Tripal job.
![Screenshot 2023-12-05 at 5 09 15 PM](https://github.com/tripal/tripal/assets/1566301/0e7ec001-39b3-40d4-9d15-7f1871d75afa)

5. Run the Tripal job on the command line and ensure there are no errors. Also check that it found the number of records you imported into chado.
6. Go to Admin > Tripal > Content and check that the new entries have been added to the content listing.
7. View one of the content pages and ensure it has the values you added to chado.
8. Edit the content page and ensure there are no errors and it updates chado.
9. Make a change to chado directly using SQL.
10. Check the content page again and confirm the change does not show up automatically (as designed ✅ )
11. Run Publish again using steps 3-5. Then refresh the content page and ensure the changes now do show up.

### Supported Content Types included in the database dump linked in the GIST:

- Gene: 50
- mRNA: 75
- Organism: 6
- Contact: 7
- Analysis: 7 (two are of type Genome Assembly and 2 are of type genome annotation)
- Project: 2 (both of type Genome Project)
- Publication: 4 (1 is the null pub)
- Protocol: 3
- Study: 3
